### PR TITLE
#289: Move to Rust ResourceIndex

### PR DESCRIPTION
**Pull request by nagato-chan**

### CODE REVIEW

File: .settings/org.eclipse.buildship.core.prefs [hunk 1]
```diff
@@ -0,0 +1,13 @@
+arguments=--init-script /home/jw/.config/Code/User/globalStorage/redhat.java/1.10.0/config_linux/org.eclipse.osgi/51/0/.cp/gradle/init/init.gradle --init-script /home/jw/.config/Code/User/globalStorage/redhat.java/1.10.0/config_linux/org.eclipse.osgi/51/0/.cp/gradle/protobuf/init.gradle
```
Comment (kirillt) [inline .settings/org.eclipse.buildship.core.prefs:1]: Please remove this file and filter `.settings` in `.gitignore`.

File: .editorconfig [hunk 2]
```diff
@@ -9,7 +9,7 @@ root = true
 
 indent_style = space
 indent_size = 4
-tab_width = 4
+tab_width = 2
```
Comment (kirillt) [inline .editorconfig:12]: No need to change.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt [hunk 3]
```diff
@@ -16,142 +20,80 @@ import space.taran.arknavigator.mvp.model.repo.preview.PreviewAndThumbnail
...
 // during application lifecycle into the DAO for the case of any unexpected exit.
-class PlainResourcesIndex internal constructor(
+class PlainResourcesIndex
+internal constructor(
     private val root: Path,
     private val dao: ResourceDao,
     resources: Map<Path, ResourceMeta>
 ) : ResourcesIndex {
 
-    private val mutex = Mutex()
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt:42]: Isn't this mutex really needed?

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt [hunk 4]
```diff
@@ -16,142 +20,80 @@ import space.taran.arknavigator.mvp.model.repo.preview.PreviewAndThumbnail
...
-                dao.deletePaths(paths.map { it.toString() })
-            }
+            chunks.forEach { paths -> dao.deletePaths(paths.map { it.toString() }) }
 
             val newResources = mutableMapOf<Path, ResourceMeta>()
-            val toInsert = diff.updated + diff.added
+
+            val toInsert = diff.added
 
             val time1 = measureTimeMillis {
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt:93]: _We should compare performance before merging._

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt [hunk 5]
```diff
@@ -165,84 +107,64 @@ class PlainResourcesIndex internal constructor(
...
-                .filter { (path, meta) ->
-                    Files.getLastModifiedTime(path) > meta.modified
-                }
-                .map { (path, _) -> path }
-
-            val added = listAllFiles(root).filter { file ->
-                !metaByPath.containsKey(file)
-            }
-
+            val time2 = measureTimeMillis { providePreviews() }
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt:110]: _We should compare performance before merging._

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt [hunk 6]
```diff
@@ -165,84 +107,64 @@ class PlainResourcesIndex internal constructor(
...
         }
+    }
 
     internal suspend fun persistResources(resources: Map<Path, ResourceMeta>) =
         withContext(Dispatchers.IO) {
             Log.d(
                 RESOURCES_INDEX,
-                "persisting " +
-                    "${resources.size} resources from root $root"
+                "persisting " + "${resources.size} resources from root $root"
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt:154]: Concatenation (`+`) is redundant now.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/kind/DocumentKindFactory.kt [hunk 7]
```diff
@@ -37,11 +37,14 @@ object DocumentKindFactory : ResourceKindFactory<ResourceKind.Document> {
...
-        return ResourceKind.Document(pages)
+        if (pages != null) {
+            return ResourceKind.Document(pages.toLong())
+        }
+        return ResourceKind.Document(null)
     }
 
     override fun fromRoom(extras: Map<MetaExtraTag, String>): ResourceKind.Document =
-        ResourceKind.Document(extras[MetaExtraTag.PAGES]?.toInt())
+        ResourceKind.Document(extras[MetaExtraTag.PAGES]?.toLong())
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/kind/DocumentKindFactory.kt:47]: Why `toLong`?

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt [hunk 8]
```diff
@@ -1,22 +1,15 @@
...
-data class ResourceMeta(
-    val id: ResourceId,
-    val name: String,
-    val extension: String,
-    val modified: FileTime,
-    val size: Long,
-    val kind: ResourceKind?
-) {
-
+class ResourceMetaExtra {
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/ResourceMeta.kt:12]: This name is confusing. Should be replaced with just normal static methods/object.
"Extra" is only untyped data in `ResourceMeta`. See `ResourceWithExtra` class.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/kind/DocumentKindFactory.kt [hunk 9]
```diff
@@ -37,11 +37,14 @@ object DocumentKindFactory : ResourceKindFactory<ResourceKind.Document> {
         val totalPages = pdfRenderer.pageCount
         val pages = if (totalPages > 0) totalPages else null
 
-        return ResourceKind.Document(pages)
+        if (pages != null) {
+            return ResourceKind.Document(pages.toLong())
+        }
+        return ResourceKind.Document(null)
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/kind/DocumentKindFactory.kt:43]: If this resource can't be determined as `ResourceKind.Document` then just return `null`.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/kind/ResourceKind.kt [hunk 10]
```diff
@@ -1,32 +1,36 @@
...
-    ) : ResourceKind(KindCode.LINK)
-
-    class PlainText : ResourceKind(KindCode.PLAINTEXT)
-
-    class Archive : ResourceKind(KindCode.ARCHIVE)
-}
+import space.taran.arklib.index.ResourceKind
+
+typealias ResourceKind = ResourceKind
+// sealed class ResourceKind(val code: KindCode) {
```
Comment (kirillt) [inline app/src/main/java/space/taran/arknavigator/mvp/model/repo/kind/ResourceKind.kt:6]: _to check with arklib_

File: .settings/org.eclipse.buildship.core.prefs (hunk 1 above)
Comment (kirillt) [reply .settings/org.eclipse.buildship.core.prefs:1]: Also I would advise to give IntelliJ IDEA a shot, you may like it more than Eclipse.

File: app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt (hunk 3 above)
Comment (nagato-chan) [reply app/src/main/java/space/taran/arknavigator/mvp/model/repo/index/PlainResourcesIndex.kt:42]: There have been a mutex underlying the call there have been a mutex, so it's no need.

**Comment by kirillt** [comment]
This PR doesn't solve the task. Probably, we should take it slower and first just move storages into arklib-android.

**Pull request closed without merging**